### PR TITLE
Add `cockroach-admin` dropshot server

### PR DESCRIPTION
**Pull request by jgallagher**
The goal here is to use this server to run `cockroach node decommission` for expunged cockroach zones. For this initial PR, the only endpoint provided wraps `cockroach node status`; wrapping `decommission` is not as trivial so I figured it should go into a separate PR.

There are currently no callers of this service, but stood up an a4x2 and confirmed I could talk to it from the switch zone via `curl`:

```
root@oxz_switch:~# curl http://[fd00:1122:3344:103::3]:32222/node/status
{"all_nodes":[{"node_id":"1","address":"[fd00:1122:3344:103::3]:32221","sql_address":"[fd00:1122:3344:103::3]:32221","build":"v22.1.9-dirty","started_at":"2024-05-24T16:47:33.137256Z","updated_at":"2024-05-24T19:01:11.345263Z","locality":"","is_available":true,"is_live":true},{"node_id":"2","address":"[fd00:1122:3344:102::3]:32221","sql_address":"[fd00:1122:3344:102::3]:32221","build":"v22.1.9-dirty","started_at":"2024-05-24T16:41:23.877326Z","updated_at":"2024-05-24T19:01:10.946872Z","locality":"","is_available":true,"is_live":true},{"node_id":"3","address":"[fd00:1122:3344:102::4]:32221","sql_address":"[fd00:1122:3344:102::4]:32221","build":"v22.1.9-dirty","started_at":"2024-05-24T16:41:24.020025Z","updated_at":"2024-05-24T19:01:11.112721Z","locality":"","is_available":true,"is_live":true},{"node_id":"4","address":"[fd00:1122:3344:101::4]:32221","sql_address":"[fd00:1122:3344:101::4]:32221","build":"v22.1.9-dirty","started_at":"2024-05-24T16:41:42.706769Z","updated_at":"2024-05-24T19:01:10.944673Z","locality":"","is_available":true,"is_live":true},{"node_id":"5","address":"[fd00:1122:3344:101::3]:32221","sql_address":"[fd00:1122:3344:101::3]:32221","build":"v22.1.9-dirty","started_at":"2024-05-24T16:41:43.079549Z","updated_at":"2024-05-24T19:01:11.326557Z","locality":"","is_available":true,"is_live":true}]}
```

### CODE REVIEW

File: sled-agent/src/services.rs [hunk 1]
```diff
@@ -1694,36 +1694,60 @@ impl ServiceManager {
...
                 )?;
 
                 let dns_service = Self::dns_install(info, None, &None).await?;
 
                 // Configure the CockroachDB service.
                 let cockroachdb_config = PropertyGroupBuilder::new("config")
-                    .add_property("listen_addr", "astring", listen_addr)
-                    .add_property("listen_port", "astring", listen_port)
+                    .add_property("listen_addr", "astring", crdb_listen_addr)
+                    .add_property("listen_port", "astring", crdb_listen_port)
```
Comment (smklein) [inline sled-agent/src/services.rs:1719]: Nitpick, feel free to tell me to go do this myself, but if we're passing the socketaddr directly to the `cockroach-admin` service, should we do the same for the `cockroachdb` service?

(Just feels weird to be asymmetrical with "ip + port" vs "socketaddr" here)

File: cockroach-admin/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+/// Start the dropshot server
+pub async fn start_server(
+    cockroach_cli: CockroachCli,
+    server_config: Config,
+) -> Result<Server, StartError> {
+    let (drain, registration) = slog_dtrace::with_drain(
+        server_config
+            .log
+            .to_logger("gateway")
```
Comment (smklein) [inline cockroach-admin/src/lib.rs:60]: Is "gateway" right here? Or is this a copy-paste from MGS?

File: cockroach-admin/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+camino.workspace = true
+chrono.workspace = true
+clap.workspace = true
+csv.workspace = true
+dropshot.workspace = true
+http.workspace = true
+illumos-utils.workspace = true
+omicron-common.workspace = true
+# See omicron-rpaths for more about the "pq-sys" dependency.
+pq-sys = "*"
```
Comment (smklein) [inline cockroach-admin/Cargo.toml:21]: I'm a little surprised we need this + the build.rs script. I *thought* we would have needed this to get access to a copy of postgres - this is normally necessary for Diesel, but is it also necessary to invoke the CLI?

File: cockroach-admin/src/bin/cockroach-admin.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    Run {
+        /// Path to the `cockroach` CLI
+        #[clap(long, action)]
+        path_to_cockroach_binary: Utf8PathBuf,
+
+        /// Socket address for a running cockroach server instance
+        #[clap(long, action)]
+        cockroach_address: SocketAddrV6,
+
+        /// Address on which this server should bin
```
Comment (smklein) [inline cockroach-admin/src/bin/cockroach-admin.rs:34]:
```suggestion
        /// Address on which this server should run
```

It's binning about too I suppose

File: cockroach-admin/src/cockroach_cli.rs [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    #[error("failed to invoke `cockroach {subcommand}`")]
+    InvokeCli {
+        subcommand: &'static str,
+        #[source]
+        err: io::Error,
+    },
+    #[error(transparent)]
+    ExecutionError(#[from] ExecutionError),
+    #[error(
+        "failed to parse `cockroack {subcommand}` output \
```
Comment (smklein) [inline cockroach-admin/src/cockroach_cli.rs:34]:
```suggestion
        "failed to parse `cockroach {subcommand}` output \
```

File: cockroach-admin/src/cockroach_cli.rs [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+        assert_eq!(statuses.len(), expected.len());
+        for (status, expected) in statuses.iter().zip(&expected) {
+            assert_eq!(status, expected);
+        }
+    }
+
+    // Ensure that if `cockroach node status` changes in a future CRDB version
+    // bump, we have a test that will fail to force us to check whether our
+    // current parsing is still valid.
+    #[tokio::test]
```
Comment (smklein) [inline cockroach-admin/src/cockroach_cli.rs:380]: hell yeah, was looking for this test, thrilled it's here

File: cockroach-admin/src/cockroach_cli.rs [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    async fn test_node_status_compatibility() {
+        let logctx = dev::test_setup_log(
+            "test_project_create_vpc_raw_returns_none_on_vni_exhaustion",
```
Comment (smklein) [inline cockroach-admin/src/cockroach_cli.rs:383]: I think the copy-paste got you here

**Review by smklein — APPROVED** [review]

File: cockroach-admin/Cargo.toml (hunk 3 above)
Comment (jgallagher) [reply cockroach-admin/Cargo.toml:21]: I guess so - I didn't really look into it when I saw the rpath failure in CI.

File: sled-agent/src/services.rs (hunk 1 above)
Comment (jgallagher) [reply sled-agent/src/services.rs]: Fixed in 7e3ca847e (after doing a bit of rework in 312e11d49 to clarify where we want an IP vs a socket address)

**Pull request merged**